File: solutions/Java/MaxSubArray_Kadane.java
public class MaxSubArray_Kadane {
    /**
     * Kadane's Algorithm - Find the maximum sum of any contiguous subarray.
     *
     * Time Complexity: O(n) - Single pass through the array
     * Space Complexity: O(1) - Only tracking current and max sums
     *
     * @param nums Array of integers
     * @return Maximum subarray sum
     */
    public static int maxSubArray(int[] nums) {
        int maxSum = nums[0];
        int currentSum = nums[0];

        for (int i = 1; i < nums.length; i++) {
            // Decide whether to extend the existing subarray or start fresh
            currentSum = Math.max(nums[i], currentSum + nums[i]);
            // Update overall maximum
            maxSum = Math.max(maxSum, currentSum);
        }

        return maxSum;
    }

    public static void main(String[] args) {
        int[] test1 = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
        System.out.println("Test 1: " + maxSubArray(test1)); // Expected: 6
    }
}

File: solutions/C++/max_subarray_kadane.cpp
#include <vector>
#include <algorithm>

using namespace std;

int maxSubArray(vector<int>& nums) {
    /**
     * Kadane's Algorithm - Find the maximum sum of any contiguous subarray.
     *
     * Time Complexity: O(n) - Single pass through the array
     * Space Complexity: O(1) - Only tracking current and max sums
     *
     * @param nums Vector of integers
     * @return Maximum subarray sum
     */
    int max_sum = nums[0];
    int current_sum = nums[0];

    for (int i = 1; i < nums.size(); i++) {
        // Decide whether to extend or start fresh
        current_sum = max(nums[i], current_sum + nums[i]);
        // Update overall maximum
        max_sum = max(max_sum, current_sum);
    }

    return max_sum;
}
